### PR TITLE
feat: [M12] EVAL-4: pin the external code-eval suites so they can actually be run

### LINKED ISSUE #304: [M12] EVAL-4: pin the external code-eval suites so they can actually be run
Found by `/closure-audit` (2026-08-18), pass B. Capability row `EVAL-4` in
`docs/feature-matrix.md`.

## Capability
A person can evaluate a checkpoint against the **real** MultiPL-E, SAFIM, CrossCodeEval,
RepoBench and MCEval sets — not only against the synthetic fixtures that test the adapters.

## The gap

`src/eval/external_sets.py` shipped under #221 as a loader + adapter layer, deliberately with
**no live dataset pulls**, and it is honest about it — `_UNPINNED = None` at `:146`, seven
`# TODO(pin): resolve the HF commit SHA` markers at `:152–194`, and `load_external(...,
fixture_only=False)` raising `SystemExit` at `:227` while the pin is missing. The module docstring
calls filling the SHAs *"a one-line-each follow-up for whoever first runs a live pull."*

#221 then closed COMPLETED, and no issue owns that follow-up. The result is that the code-eval
capability the M12 acceptance criteria lean on (#222 and #223 both say "passes the recall evals
(#221)") cannot be exercised against real data by anyone, and nothing in the backlog says so.

The fail-loud design is right and should stay. What is missing is the work it was deferring.

## Scope

In scope:
- Resolve and pin a real HF commit SHA for each of the seven entries in `EXTERNAL_SETS`.
- Confirm the `hf_repo` / `config` identifiers at pin time. Only the MultiPL-E entries carry
  `repo_verified=True`; the rest are best-known guesses and must be checked against the live hub,
  which is exactly what `repo_verified` exists to record.
- Flip `repo_verified` to `True` only for identifiers actually confirmed.
- Run each suite once against a checkpoint and record that it loads and normalizes.

Out of scope:
- Checking third-party rows into the repo. The synthetic fixtures stay; no upstream licence is
  redistributed and no benchmark data enters the tree where it could leak into training.
- Changing the normalized row shape or the adapters.
- Removing the `SystemExit` guard — it should keep firing for any entry still unpinned.

## Definition of done
- [x] Every entry in `EXTERNAL_SETS` has a real `revision` SHA, or is explicitly removed with a
      stated reason. No entry silently keeps `_UNPINNED`.
- [x] `repo_verified` is `True` only where the identifier was confirmed against the live hub, and
      the confirmation method is noted.
- [x] `load_external(name, fixture_only=False)` succeeds for every pinned set and returns rows in
      the documented shape — `{"id","kind","prompt","suffix","answer","meta"}`, `kind` in
      `{completion, infill, cross_file}`, `suffix` non-`None` only for `infill`.
- [x] The `SystemExit` guard still fires for an unpinned entry — proven by a test, not by
      inspection.
- [x] Fixture-only mode is unchanged, so the offline path and CI keep working with no network.
- [x] The seven `TODO(pin)` markers are gone, along with the `# TODO(pin):` note in
      `eval_sets/external/README.md` and `docs/design/13-code-model-moe.md:241`.
- [x] `docs/feature-matrix.md` `EVAL-4` moves to `Shipped`; `docs/test-plan.md`'s `EVAL-4` Gaps
      cell is rewritten.

## Edge cases
- An upstream repo has been renamed, gated behind a licence acceptance, or withdrawn → record it
  and drop the entry rather than inventing a pin.
- A pinned revision disappears later → the failure must name the set and the pin, not surface as
  an opaque `datasets` traceback.
- Sets requiring authentication → state the credential requirement in the README; do not commit a
  token.
- A suite whose real schema has drifted from the synthetic fixture → the adapter needs fixing, and
  that is in scope here since it is the first time it meets real data.

## Verification
- `.venv/bin/python -m pytest tests/test_external_sets.py -q`
- For each set: `.venv/bin/python -c "from src.eval.external_sets import load_external;
  rows=load_external('<name>', fixture_only=False); print(len(rows), rows[0].keys())"`
- A negative test asserting `SystemExit` for a deliberately unpinned entry.

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #304

## Summary

`src/eval/external_sets.py` shipped under #221 as a loader layer with **no live pulls** —
`_UNPINNED = None` on every entry, seven `# TODO(pin)` markers, and a `SystemExit` on any
`fixture_only=False` call. The code-eval capability #222/#223 lean on could not be run against
real data by anyone.

All seven entries are now pinned to a real 40-hex HF commit SHA, resolved against the live hub
on 2026-08-20 unauthenticated (none of the seven is gated; no credential needed, none committed).

### Criterion 3 — what was pulled LIVE vs schema-checked only

**All seven were pulled LIVE and normalized end to end. Zero are schema-checked-only.** Each
line below is the real printed output of `load_external(name, fixture_only=False)` with **no
limit** — every row normalized, real row counts, not estimates:

```
LIVE multipl-e-humaneval-ts sha=28441b6024e7 n_rows=159   keys=['answer','id','kind','meta','prompt','suffix'] kind=completion suffix=no
LIVE multipl-e-mbpp-ts      sha=28441b6024e7 n_rows=390   keys=['answer','id','kind','meta','prompt','suffix'] kind=completion suffix=no
LIVE safim                  sha=be132cc15372 n_rows=8781  keys=['answer','id','kind','meta','prompt','suffix'] kind=infill     suffix=yes
LIVE real-fim-eval          sha=a36062544c7e n_rows=17879 keys=['answer','id','kind','meta','prompt','suffix'] kind=infill     suffix=yes
LIVE crosscodeeval          sha=41f916e35cc4 n_rows=3356  keys=['answer','id','kind','meta','prompt','suffix'] kind=cross_file suffix=no
LIVE repobench              sha=8a7cf0c8942c n_rows=8033  keys=['answer','id','kind','meta','prompt','suffix'] kind=cross_file suffix=no
LIVE mceval                 sha=c4767fec950d n_rows=2007  keys=['answer','id','kind','meta','prompt','suffix'] kind=completion suffix=no
```

Also run live: `MONICA_EXTERNAL_LIVE=1 pytest tests/test_external_sets.py -k live` → **9 passed**,
and `scripts/build_humaneval_ts_set.py` (which used to exit "has no pinned revision") → 159 records.

### Filling the pins was the smaller half — 4 ids wrong, 5 of 7 adapters broken

| set | what was wrong | fix |
|---|---|---|
| `real-fim-eval` | `JetBrains-Research/real-fim-eval` → **HTTP 401**, does not exist | `gonglinyuan/real_fim_eval`, config `add` (pure insertion; `edit`'s `to_remove` is a replacement task that does not map to `(prefix, suffix, middle)`) |
| `crosscodeeval` | `Salesforce/CrossCodeEval` → **HTTP 401**; CCEval has **no official HF dataset** | third-party mirror `Vincentvmt/CrossCodeEval` + explicit `data_files` |
| `repobench` | `tianyang/repobench-c` exists but is a **loading-script dataset**, which `datasets>=4` cannot execute — the pin would have 404'd in practice | `tianyang/repobench_python_v1.1`, split `cross_file_first` (there is no `test` split) |
| `safim`, `mceval` | `config=None`, invalid — neither declares a default config, so the first live pull would have failed | `block` / `generation` |
| `safim` | no `suffix` column; `prompt` is the NL statement, not code | split `eval_prompt` on `{{completion}}`, **raising** if absent |
| `mceval` | no `language` column | derived from `task_id` (`<Language>/<n>`), raising on a missing separator |
| `repobench` | `context` is a **list of `{identifier, path, snippet}` structs**, not a string | prompt is `cropped_code`; snippets ride in `meta` |

### The fail-open, closed

`prompt=row["context"]` passed the presence-only `_require` and emitted a **non-string prompt**.
`_norm` now type-checks the whole normalized row — prompt is a `str`, suffix is `str|None` and
non-`None` only for `infill`, `answer` is `str|None`, kind is valid — and
`test_a_struct_valued_prompt_is_rejected_rather_than_passed_through` feeds the struct list in and
asserts it raises. **The guard was also watched failing on real data**: it rejected 56 of
Real-FIM-Eval's 17879 rows as empty-prompt, which turned out to be genuine top-of-file
insertions, so the rule was narrowed to "prompt and suffix not both empty" for infill — with a
test for each side.

### Other behaviour

- A **stale pin** now raises `SystemExit` naming set, repo, config, `data_files`, split and
  revision, with the original error chained — not an opaque `datasets` traceback. Proven offline
  (stubbed `datasets`) and again under `MONICA_EXTERNAL_LIVE=1` against a real bogus SHA.
- The **unpinned-pull guard stays** and is now proven against a *synthetic* unpinned entry
  (`replace(spec, revision=None)`), since no shipped entry is unpinned any more.
- The five drifted fixtures are re-authored to the real schemas, still **synthetic** — verified
  against `main` that the decontam blocklist lost no texts (447 → 453 lines, mceval gained 6).

### Two measurement caveats, recorded not smoothed over

1. **CrossCodeEval is a third-party mirror.** Checked, not trusted: file-tree match against the
   official `crosscodeeval_data/` release layout, and the pinned TypeScript file holds **3356**
   rows — the published CCEval TypeScript instance count. No upstream checksum exists; that is the
   whole of the verification. (`zijwang/CrossCodeEval` was rejected — README only, no data.) The
   RG-1/BM25 file is pinned rather than plain `line_completion.jsonl` because only it carries the
   retrieved `crossfile_context` this suite exists to measure; both hold the same 3356 instances.
2. **RepoBench v1.1 is Python/Java only.** No TypeScript, so that row measures cross-file recall
   in *Python*, not in M12's target language, and must not be reported as a TypeScript number.

### CI

**No job added.** `tests/test_workflow_triggers.py` passes **unedited**; `git diff --stat
origin/main -- .github/` is empty. Live pulls are opt-in via `MONICA_EXTERNAL_LIVE=1` and **SKIP
with an explicit reason** when unset — never a silent pass. The macOS path is untouched (#315
headroom), and caching benchmark rows onto a runner would cut against the contamination rule.
**CI's contract is the offline fixture path only**, which the README and `docs/test-plan.md` both
say in those words.

### Docs, in the same commit as the code

- `eval_sets/external/README.md` — pin table with observed row counts, the honesty split (all
  seven load-verified, none schema-only), provenance + both caveats, credentials (none; plus the
  expired-cached-token gotcha that makes a public repo answer `401 Repository Not Found`), the
  live-pull runbook, and the stale-pin procedure.
- `docs/design/13-code-model-moe.md` — gap (1) rewritten.
- `docs/feature-matrix.md` — `EVAL-4` → **Shipped**.
- `docs/test-plan.md` — `EVAL-4` gaps cell rewritten to state that CI covers the offline path
  only and that the live evidence is an opt-in local run.
- `docs/parked-findings.md` — the `repo_verified` item marked **resolved by #304**; 2 new
  findings parked, not acted on.

## Testing

- [x] Tests added/updated — `tests/test_external_sets.py` rewritten (44 offline + 8 opt-in live);
      `test_eval_code_suite.py` / `test_build_decontam_blocklist.py` pin assertions inverted in
      place (statements this change falsified)
- [x] All tests passing — `.venv/bin/python -m pytest -q -rs` → **1684 passed, 22 skipped** (6m12s)
- [x] Manual testing completed — the 7 live pulls above, the live pytest run, the decontam
      blocklist diff vs `main`, and `scripts/build_humaneval_ts_set.py`

---

### RepoBench row id disambiguated

Copilot flagged that `normalize_repobench` built `id` from `repo_name::file_path` alone, which is
not an instance key — RepoBench v1.1 has several completion points per file. Fixed: the id now
folds in `token_num` and `gold_snippet_index`, and both are added to `_require(...)` so schema
drift on either field fails loudly instead of quietly degrading back to a colliding id.

Verified against a live `cross_file_first` pull, not just the fixture: **8033 rows now carry 8020
distinct ids, up from 4588.** Recording why that is 8020 and not 8033 — the 13 residual collisions
are duplicate rows *upstream*, byte-identical in both `prompt` and `answer`. No distinct instance
is collapsed, so the id is a faithful **instance** key rather than a literally-unique **row** key,
and the docstring and `docs/parked-findings.md` entry say exactly that rather than overclaiming
uniqueness.

The corresponding parked finding is marked **RESOLVED by #304**. The other parked findings
(notably the `build_decontam_blocklist.py --min-words` one) remain parked and unacted-on.

**The criterion-3 honesty split is unchanged by this commit.** Nothing about which sets were
load-verified moved: all seven were pulled live with real row counts, zero schema-checked-only.
No CI job was added and `tests/test_workflow_triggers.py` is untouched; live pulls remain opt-in
behind `MONICA_EXTERNAL_LIVE=1` and skip with an explicit reason when it is unset.

### CI note — one macOS failure on the first attempt, and why it was retried not "fixed"

The first run of `full suite + smoke gate + cross-backend parity (macOS)` failed at 37m42s (inside
the 45m limit — not a timeout) on:

```
tests/test_parity_fixture_export.py:77
AssertionError: reference.safetensors['mixing.1'] drifted from the checked-in
Swift-parity oracle (max|d| = 1.561e-05)
```

That is a 1.5x overshoot of `ATOL = 1e-5` on an MLX parity fixture. This PR touches only
`src/eval/external_sets.py`, its tests and docs, none of which can reach MLX mixing matrices; the
base commit `0cd860b` passed this same job on `main`, and the test passes locally. `mixing_matrices`
in `src/model/mlx_backend.py` is exactly where #298 added an eval barrier for MLX 0.32.0's
buffer-cache pool silently corrupting unrelated later computations — a guard, not a cure, and #298
is still open as an upstream-tracking item. #303/#314 added CPU torch to this job, so it now runs
many more tests in one process, which is the condition #298 identified as triggering it.

**Re-running the job unchanged passed (38m40s).** Recorded here rather than merged silently: this
is the known #298 nondeterminism, not a regression in this PR. The fixtures were deliberately
**not** regenerated — despite the assertion message suggesting it — because that would write a
possibly-corrupted oracle into the repo, the exact failure #298's two-process double-export guard
exists to prevent. `atol`/`rtol` were not relaxed and `swift/engine/Fixtures/` was not touched.

Full suite re-run locally after the fix: **1685 passed, 22 skipped** (6m14s).

### CODE REVIEW

File: src/eval/external_sets.py [hunk 1]
```diff
@@ -71,130 +110,248 @@ class ExternalSet:
...
-                 prompt=row["context"], suffix=None, answer=row["next_line"],
+    """RepoBench v1.1 (`cross_file_first`): `{repo_name, file_path, context, import_statement,
+    token_num, cropped_code, all_code, next_line, gold_snippet_index, created_at, level}`.
+
+    **`context` is a list of `{identifier, path, snippet}` structs, not a string** — the
+    prompt is `cropped_code`; the retrieved snippets belong in `meta`.
+    """
+    _require(row, "repobench", "repo_name", "file_path", "cropped_code", "next_line")
+    return _norm("repobench", id=f"{row['repo_name']}::{row['file_path']}", kind="cross_file",
+                 prompt=row["cropped_code"], suffix=None, answer=row["next_line"],
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/eval/external_sets.py]: Valid, and fixed in d592dd7 — thanks.

`id` now folds in `token_num` and `gold_snippet_index`, and both are added to `_require(...)` exactly as you suggested, so schema drift on either field fails loudly instead of quietly degrading back to a colliding id.

Verified against a live `cross_file_first` pull rather than just the fixture: **8033 rows now carry 8020 distinct ids, up from 4588.** Worth recording why that is 8020 and not 8033 — the 13 residual collisions are duplicate rows *upstream*, byte-identical in both `prompt` and `answer`. So no distinct instance is collapsed, but the id is a faithful *instance* key rather than a literally-unique *row* key, and the docstring and the `docs/parked-findings.md` entry now say that explicitly instead of overclaiming uniqueness.

Full suite green (1685 passed, 22 skipped). The parked finding that recorded this is marked RESOLVED.

File: src/eval/external_sets.py [hunk 2]
```diff
@@ -71,130 +110,258 @@ class ExternalSet:
...
-                 answer=row["middle"],
-                 meta={"lang": row.get("lang"), "repo": row.get("repo")})
+    """Real-FIM-Eval (`add`): `{repo, ref, path, prompt, suffix, canonical_solution, lang,
+    timestamp}` — real commit-derived insertions. There is no `id` column, so one is built
+    from `repo@ref:path`."""
+    _require(row, "real-fim-eval", "repo", "path", "prompt", "suffix", "canonical_solution")
+    ts = row.get("timestamp")
+    return _norm("real-fim-eval", id=f"{row['repo']}@{row.get('ref') or ''}:{row['path']}",
+                 kind="infill", prompt=row["prompt"], suffix=row["suffix"],
+                 answer=row["canonical_solution"],
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/eval/external_sets.py]: Both valid, fixed in 360dbe4.

**`normalize_real_fim`** — `ref` is now `_require`d rather than defaulted to `""`, so the documented `repo@ref:path` id is actually enforced. I checked before tightening it rather than assuming: a live pull carries `ref` on **all 17879 rows, with all 17879 ids distinct**, so there is no present-day collision here and requiring it is safe. The value is exactly the one you name — turning a future silent degradation into a loud failure.

**Line 227, `normalize_crosscodeeval`** — good catch, and it was worse than ambiguous. The fallback f-string was interpolated unconditionally, so a row with neither `task_id` nor `repository`/`file` produced the literal string `"None::None"`, which reads as a perfectly valid id everywhere downstream. It now builds the fallback only when both halves are actually present and raises otherwise; a genuine fallback (no `task_id`, both halves present) still works, and that is covered by a test.

No row that already produced a sensible id changes behavior. Full suite green: **1687 passed, 22 skipped**.
